### ISSUE TITLE
Some improvement suggestions

### ISSUE THREAD

**Issue by timreichen** [postcss/postcss#1503]
I like postcss so much that I am using it as a base for all css stuff in my [bundler](https://github.com/timreichen/Bundler) project, which is used as a front-end bundler run on [deno](https://deno.land).
So I was looking for ways to make the process more efficient and gathered a list of possible enhancement as well as tweaks to align the synstax more with js conventions.

1. Process AST directly
The `process` method takes a string that is parsed to an AST. I think we should be able to run plugins by passing AST instead of a string. For various reasons I need to run postcss process twice, it would be neat to be able to reuse the AST instead of parsing twice.

2. `postcss.stringify(…)` instead of `ast.toString(…)`
    Since `postcss.parse(string, options={}) -> ast` is a thing, would it make sense to have `postcss.stringify(ast, options={}) -> string`. `ast.toString(…)` seems a bit inconsistent compared to the js convention like in the JSON namespace.

3. parse function opts default
    Not a problem in js, but in ts: In parser.js `function parse (css, opts) {…}` requires an opts parameter, which is passed to  `new Input(css, opts)`, in which `opts={}`. So in ts `postcss.parse("div { apprearance: none }", {})` needs an opts, even if it is empty. If  `function parse (css, opts={}) {…}` that wouldn't be the case and align more with the js convention..

4. using plugins and plugin class
    Apparently you can set the postcss plugins in different ways:

    ```ts
    const processor = postcss([plugin1, plugin2])
    ```
    ```ts
    processor.plugins.push(plugin1, plugin2)
    ```
    ```ts
    processor
    .use(plugin1)
    .use(plugin2)
    ```

    I wonder if it would be better to drop the `use` method in favour of plugins.push, since it is more descriptive? I guess the plugins must be normalized for some reason, but if plugins were some kind of Plugin class instances, wouldn't that make the normalization obsolete?
5. esm modules
    usage of esm instead of cjs since the feature is stable since node 15.
6. Processor class
    Be able to `const processor = new Processor()` instead of `postcss()`. Would be nice if `Processor` was exposed instead of the postcss function, since it is just a convenience function.

I am aware that some of these suggestions would be breaking changes, but I think it would bring the project to the next level. What are your thoughts?

**Comment by ai** [comment]
> 1. Process AST directly

You already can send AST to `postcss().process`:

```js
let root = postcss.parse(css)
postcss().process(root)
```

> 2. postcss.stringify(…) instead of ast.toString(…)

I can’t change the essential API.

> 3. parse function opts default

You need to set `from` option to generate the correct source map

> 4. using plugins and plugin class

PostCSS is too popular to remove methods without a good reason

> 5. esm modules

PostCSS 8 already supports ESM. Look into the npm package.

> 6. Processor class

Why do you need to add an additional API to create `Processor` instance?

**Comment by timreichen** [comment]
Thank you for your fast answer!
1. Fantastic, I somehow missed that!
2. understandable.
3. Maybe then it should be declared mandatory instead of optional in d.ts? Because at the moment an empty opts does not raise any error.
4. also understandable. I guess I this would be an unexpected breaking change, certainly not for a minor update. I thinl this would be more a like a feature for a future iteration or some kind of successor of postcss. Is there something like this planned?
5. Right, awesome!
6. I think calling `new` on a class in code is much more expressive than calling a function that creates an instance. And since that is going on under the hood anyway, why not expose it?
    ```ts
    function postcss (...plugins) {
      if (plugins.length === 1 && Array.isArray(plugins[0])) {
        plugins = plugins[0]
      }
      return new Processor(plugins, postcss)
    }
    ```

**Comment by ai** [comment]
> Maybe then it should be declared mandatory instead of optional in d.ts? Because at the moment an empty opts does not raise any error.

Yeap. Good idea.

> I think this would be more a like a feature for a future iteration or some kind of successor of postcss. Is there something like this planned?

Am I right that we are talking about dropping the `use` method? It is a major change, not a feature.

> And since that is going on under the hood anyway, why not expose it? 

I do not want to maintain more APIs. But if you have a case for using `new Processor()` we can add it to type definitions.

**Comment by timreichen** [comment]
> Am I right that we are talking about dropping the use method? It is a major change, not a feature.

Yeah, change not feature. _I think this would be more a like a change for a future iteration or some kind of successor of postcss. Is there something like this planned?_

> I do not want to maintain more APIs. But if you have a case for using new Processor() we can add it to type definitions.

The classes `Processor`, `Stringifier`, `Parser` etc. are already part of postcss, so I meant to expose these classes in `postcss.js`  as an export so they can be imported like `import { Processor } from "postcss"` for example. At the moment the only way is to of call a creator function like `postcss()`, `parse()` or `stringify()` which basically create an instance but also do some other stuff or to import the respective files (`import { Processor } from "postcss/processor"`) which I think is not quite as an elegant solution.

**Comment by ai** [comment]
> Is there something like this planned?

Nope. CSS tooling is not a great market to spend resources 😓.

> I think is not quite as an elegant solution.

The problem is that you still need `postcss` to create `Processor` instance:

```diff
- import postcss from 'postcss'
- postcss(plugins)
+ import postcss, { Processor } from 'postcss'
+ new Processor(plugins, postcss)
```

It looks ugly (for me) in this way.

**Comment by timreichen** [comment]
> Nope. CSS tooling is not a great market to spend resources 😓.

I get you, that is a shame though😶😇

> The problem is that you still need postcss to create Processor instance:

But in the Processor constructor there is no second argument. Or did I miss something?
```ts
// lib/processor.js
class Processor {
  constructor (plugins = []) {
    this.version = '8.2.3'
    this.plugins = this.normalize(plugins)
  }
```

**Comment by ai** [comment]
> But in the Processor constructor there is no second argument. Or did I miss something?

Hm. Seems like it was a mistake in `lib/postcss.js`. Fixed d2845ed

Send PR with exporting `Processor` class.

**Comment by ai** [comment]
We will exported `Processor` in next 8.4.
